### Will a Customer Accept the Coupon?

**Context**

Imagine driving through town and a coupon is delivered to your cell phone for a restaraunt near where you are driving. Would you accept that coupon and take a short detour to the restaraunt? Would you accept the coupon but use it on a subsequent trip? Would you ignore the coupon entirely? What if the coupon was for a bar instead of a restaraunt? What about a coffee house? Would you accept a bar coupon with a minor passenger in the car? What about if it was just you and your partner in the car? Would weather impact the rate of acceptance? What about the time of day?

Obviously, proximity to the business is a factor on whether the coupon is delivered to the driver or not, but what are the factors that determine whether a driver accepts the coupon once it is delivered to them? How would you determine whether a driver is likely to accept a coupon?

**Overview**

The goal of this project is to use what you know about visualizations and probability distributions to distinguish between customers who accepted a driving coupon versus those that did not.

**Data**

This data comes to us from the UCI Machine Learning repository and was collected via a survey on Amazon Mechanical Turk. The survey describes different driving scenarios including the destination, current time, weather, passenger, etc., and then ask the person whether he will accept the coupon if he is the driver. Answers that the user will drive there ‘right away’ or ‘later before the coupon expires’ are labeled as ‘Y = 1’ and answers ‘no, I do not want the coupon’ are labeled as ‘Y = 0’.  There are five different types of coupons -- less expensive restaurants (under \\$20), coffee houses, carry out & take away, bar, and more expensive restaurants (\\$20 - \\$50). 

**Deliverables**

Your final product should be a brief report that highlights the differences between customers who did and did not accept the coupons.  To explore the data you will utilize your knowledge of plotting, statistical summaries, and visualization using Python. You will publish your findings in a public facing github repository as your first portfolio piece. 





### Data Description
Keep in mind that these values mentioned below are average values.

The attributes of this data set include:
1. User attributes
    -  Gender: male, female
    -  Age: below 21, 21 to 25, 26 to 30, etc.
    -  Marital Status: single, married partner, unmarried partner, or widowed
    -  Number of children: 0, 1, or more than 1
    -  Education: high school, bachelors degree, associates degree, or graduate degree
    -  Occupation: architecture & engineering, business & financial, etc.
    -  Annual income: less than \\$12500, \\$12500 - \\$24999, \\$25000 - \\$37499, etc.
    -  Number of times that he/she goes to a bar: 0, less than 1, 1 to 3, 4 to 8 or greater than 8
    -  Number of times that he/she buys takeaway food: 0, less than 1, 1 to 3, 4 to 8 or greater
    than 8
    -  Number of times that he/she goes to a coffee house: 0, less than 1, 1 to 3, 4 to 8 or
    greater than 8
    -  Number of times that he/she eats at a restaurant with average expense less than \\$20 per
    person: 0, less than 1, 1 to 3, 4 to 8 or greater than 8
    -  Number of times that he/she goes to a bar: 0, less than 1, 1 to 3, 4 to 8 or greater than 8
    

2. Contextual attributes
    - Driving destination: home, work, or no urgent destination
    - Location of user, coupon and destination: we provide a map to show the geographical
    location of the user, destination, and the venue, and we mark the distance between each
    two places with time of driving. The user can see whether the venue is in the same
    direction as the destination.
    - Weather: sunny, rainy, or snowy
    - Temperature: 30F, 55F, or 80F
    - Time: 10AM, 2PM, or 6PM
    - Passenger: alone, partner, kid(s), or friend(s)


3. Coupon attributes
    - time before it expires: 2 hours or one day

In [440]:
import matplotlib.pyplot as plt
import seaborn as sns
import pandas as pd
import numpy as np
import plotly.express as px

### Problems

Use the prompts below to get started with your data analysis.  

1. Read in the `coupons.csv` file.




In [441]:
data = pd.read_csv('data/coupons.csv')

In [442]:
data.head()

destination  passanger weather  temperature  time  \
0  No Urgent Place      Alone   Sunny           55   2PM   
1  No Urgent Place  Friend(s)   Sunny           80  10AM   
2  No Urgent Place  Friend(s)   Sunny           80  10AM   
3  No Urgent Place  Friend(s)   Sunny           80   2PM   
4  No Urgent Place  Friend(s)   Sunny           80   2PM   

                  coupon expiration  gender age      maritalStatus  ...  \
0        Restaurant(<20)         1d  Female  21  Unmarried partner  ...   
1           Coffee House         2h  Female  21  Unmarried partner  ...   
2  Carry out & Take away         2h  Female  21  Unmarried partner  ...   
3           Coffee House         2h  Female  21  Unmarried partner  ...   
4           Coffee House         1d  Female  21  Unmarried partner  ...   

   CoffeeHouse CarryAway RestaurantLessThan20 Restaurant20To50  \
0        never       NaN                  4~8              1~3   
1        never       NaN                  4~8              1~3   
2        never       NaN                  4~8              1~3   
3        never       NaN                  4~8              1~3   
4        never       NaN                  4~8              1~3   

  toCoupon_GEQ5min toCoupon_GEQ15min toCoupon_GEQ25min direction_same  \
0                1                 0                 0              0   
1                1                 0                 0              0   
2                1                 1                 0              0   
3                1                 1                 0              0   
4                1                 1                 0              0   

  direction_opp  Y  
0             1  1  
1             1  0  
2             1  1  
3             1  0  
4             1  0  

[5 rows x 26 columns]

In [443]:
data.count()

destination             12684
passanger               12684
weather                 12684
temperature             12684
time                    12684
coupon                  12684
expiration              12684
gender                  12684
age                     12684
maritalStatus           12684
has_children            12684
education               12684
occupation              12684
income                  12684
car                       108
Bar                     12577
CoffeeHouse             12467
CarryAway               12533
RestaurantLessThan20    12554
Restaurant20To50        12495
toCoupon_GEQ5min        12684
toCoupon_GEQ15min       12684
toCoupon_GEQ25min       12684
direction_same          12684
direction_opp           12684
Y                       12684
dtype: int64

2. Investigate the dataset for missing or problematic data.

In [444]:
# Check which columns contain NaN values
data.isnull().any()

destination             False
passanger               False
weather                 False
temperature             False
time                    False
coupon                  False
expiration              False
gender                  False
age                     False
maritalStatus           False
has_children            False
education               False
occupation              False
income                  False
car                      True
Bar                      True
CoffeeHouse              True
CarryAway                True
RestaurantLessThan20     True
Restaurant20To50         True
toCoupon_GEQ5min        False
toCoupon_GEQ15min       False
toCoupon_GEQ25min       False
direction_same          False
direction_opp           False
Y                       False
dtype: bool

In [445]:
# For each column with NaN, check unique values to get a sense of the data
data.car.unique()

array([nan, 'Scooter and motorcycle', 'crossover', 'Mazda5',
       'do not drive', 'Car that is too old to install Onstar :D'],
      dtype=object)

Since the scope of the study is to check the effects of coupon on drivers, we will remove rows with 'do not drive' in car column. There are only 108 values in car column and the rest in NaN. We will replace NaN with unknown car to preserve the records.

In [446]:
data.Bar.unique()

array(['never', 'less1', '1~3', 'gt8', nan, '4~8'], dtype=object)

We'll assume that nan, never, and less1 are respondents that don't go to bars

In [447]:
data.CoffeeHouse.unique()

array(['never', 'less1', '4~8', '1~3', 'gt8', nan], dtype=object)

Just like in Bar, nan, never, and less1 will be equal to 0. The same transformation will be applied to CarryAway, RestaurantLessThan20, Restaurant20To50

In [448]:
data.CarryAway.unique()

array([nan, '4~8', '1~3', 'gt8', 'less1', 'never'], dtype=object)

In [449]:
data.RestaurantLessThan20.unique()

array(['4~8', '1~3', 'less1', 'gt8', nan, 'never'], dtype=object)

In [450]:
data.Restaurant20To50.unique()

array(['1~3', 'less1', 'never', 'gt8', '4~8', nan], dtype=object)

In [451]:
# how check Bar column for coupon = 'Bar'
data.query("coupon == 'Bar'")[['Bar','Y']].groupby('Bar').size()

Bar
1~3      397
4~8      150
gt8       49
less1    570
never    830
dtype: int64

In [477]:
data.maritalStatus.unique()

array(['Unmarried partner', 'Single', 'Married partner', 'Divorced',
       'Widowed'], dtype=object)

3. Decide what to do about your missing data -- drop, replace, other...

In [452]:
# replace rows with NaN in car with unknown and assign to cleandata
cleandata = data.fillna('unknown')

In [453]:
# remove entries where car = 'do not drive'
cleandata = cleandata.query("car != 'do not drive'")

In [454]:
# perform transformations on Bar, CoffeeHouse, CarryAway, RestaurantLessThan20, Restaurant20To50 and convert to numeric values 

cleandata['Bar'] = cleandata['Bar'].fillna(0)
cleandata['CoffeeHouse'] = cleandata['CoffeeHouse'].fillna(0)
cleandata['RestaurantLessThan20'] = cleandata['RestaurantLessThan20'].fillna(0)
cleandata['Restaurant20To50'] = cleandata['Restaurant20To50'].fillna(0)

conv_to_numeric = {'Bar': {'never': 0, 'less1': 0, '1~3': 1, '4~8': 4, 'gt8': 8},
                    'CoffeeHouse': {'never': 0, 'less1': 0, '1~3': 1, '4~8': 4, 'gt8': 8},
                    'CarryAway': {'never': 0, 'less1': 0, '1~3': 1, '4~8': 4, 'gt8': 8},
                    'RestaurantLessThan20': {'never': 0, 'less1': 0, '1~3': 1, '4~8': 4, 'gt8': 8},
                    'Restaurant20To50': {'never': 0, 'less1': 0, '1~3': 1, '4~8': 4, 'gt8': 8}}

newdata=cleandata.replace(conv_to_numeric)

In [455]:
# check # of rows where car = "do not drive"
newdata

destination  passanger weather  temperature  time  \
0      No Urgent Place      Alone   Sunny           55   2PM   
1      No Urgent Place  Friend(s)   Sunny           80  10AM   
2      No Urgent Place  Friend(s)   Sunny           80  10AM   
3      No Urgent Place  Friend(s)   Sunny           80   2PM   
4      No Urgent Place  Friend(s)   Sunny           80   2PM   
...                ...        ...     ...          ...   ...   
12679             Home    Partner   Rainy           55   6PM   
12680             Work      Alone   Rainy           55   7AM   
12681             Work      Alone   Snowy           30   7AM   
12682             Work      Alone   Snowy           30   7AM   
12683             Work      Alone   Sunny           80   7AM   

                      coupon expiration  gender age      maritalStatus  ...  \
0            Restaurant(<20)         1d  Female  21  Unmarried partner  ...   
1               Coffee House         2h  Female  21  Unmarried partner  ...   
2      Carry out & Take away         2h  Female  21  Unmarried partner  ...   
3               Coffee House         2h  Female  21  Unmarried partner  ...   
4               Coffee House         1d  Female  21  Unmarried partner  ...   
...                      ...        ...     ...  ..                ...  ...   
12679  Carry out & Take away         1d    Male  26             Single  ...   
12680  Carry out & Take away         1d    Male  26             Single  ...   
12681           Coffee House         1d    Male  26             Single  ...   
12682                    Bar         1d    Male  26             Single  ...   
12683      Restaurant(20-50)         2h    Male  26             Single  ...   

       CoffeeHouse CarryAway RestaurantLessThan20 Restaurant20To50  \
0                0   unknown                    4                1   
1                0   unknown                    4                1   
2                0   unknown                    4                1   
3                0   unknown                    4                1   
4                0   unknown                    4                1   
...            ...       ...                  ...              ...   
12679            0         1                    4                1   
12680            0         1                    4                1   
12681            0         1                    4                1   
12682            0         1                    4                1   
12683            0         1                    4                1   

      toCoupon_GEQ5min toCoupon_GEQ15min toCoupon_GEQ25min direction_same  \
0                    1                 0                 0              0   
1                    1                 0                 0              0   
2                    1                 1                 0              0   
3                    1                 1                 0              0   
4                    1                 1                 0              0   
...                ...               ...               ...            ...   
12679                1                 0                 0              1   
12680                1                 0                 0              0   
12681                1                 0                 0              1   
12682                1                 1                 1              0   
12683                1                 0                 0              1   

      direction_opp  Y  
0                 1  1  
1                 1  0  
2                 1  1  
3                 1  0  
4                 1  0  
...             ... ..  
12679             0  1  
12680             1  1  
12681             0  0  
12682             1  0  
12683             0  0  

[12662 rows x 26 columns]

4. What proportion of the total observations chose to accept the coupon? 



In [456]:
cleandata

destination  passanger weather  temperature  time  \
0      No Urgent Place      Alone   Sunny           55   2PM   
1      No Urgent Place  Friend(s)   Sunny           80  10AM   
2      No Urgent Place  Friend(s)   Sunny           80  10AM   
3      No Urgent Place  Friend(s)   Sunny           80   2PM   
4      No Urgent Place  Friend(s)   Sunny           80   2PM   
...                ...        ...     ...          ...   ...   
12679             Home    Partner   Rainy           55   6PM   
12680             Work      Alone   Rainy           55   7AM   
12681             Work      Alone   Snowy           30   7AM   
12682             Work      Alone   Snowy           30   7AM   
12683             Work      Alone   Sunny           80   7AM   

                      coupon expiration  gender age      maritalStatus  ...  \
0            Restaurant(<20)         1d  Female  21  Unmarried partner  ...   
1               Coffee House         2h  Female  21  Unmarried partner  ...   
2      Carry out & Take away         2h  Female  21  Unmarried partner  ...   
3               Coffee House         2h  Female  21  Unmarried partner  ...   
4               Coffee House         1d  Female  21  Unmarried partner  ...   
...                      ...        ...     ...  ..                ...  ...   
12679  Carry out & Take away         1d    Male  26             Single  ...   
12680  Carry out & Take away         1d    Male  26             Single  ...   
12681           Coffee House         1d    Male  26             Single  ...   
12682                    Bar         1d    Male  26             Single  ...   
12683      Restaurant(20-50)         2h    Male  26             Single  ...   

       CoffeeHouse CarryAway RestaurantLessThan20 Restaurant20To50  \
0            never   unknown                  4~8              1~3   
1            never   unknown                  4~8              1~3   
2            never   unknown                  4~8              1~3   
3            never   unknown                  4~8              1~3   
4            never   unknown                  4~8              1~3   
...            ...       ...                  ...              ...   
12679        never       1~3                  4~8              1~3   
12680        never       1~3                  4~8              1~3   
12681        never       1~3                  4~8              1~3   
12682        never       1~3                  4~8              1~3   
12683        never       1~3                  4~8              1~3   

      toCoupon_GEQ5min toCoupon_GEQ15min toCoupon_GEQ25min direction_same  \
0                    1                 0                 0              0   
1                    1                 0                 0              0   
2                    1                 1                 0              0   
3                    1                 1                 0              0   
4                    1                 1                 0              0   
...                ...               ...               ...            ...   
12679                1                 0                 0              1   
12680                1                 0                 0              0   
12681                1                 0                 0              1   
12682                1                 1                 1              0   
12683                1                 0                 0              1   

      direction_opp  Y  
0                 1  1  
1                 1  0  
2                 1  1  
3                 1  0  
4                 1  0  
...             ... ..  
12679             0  1  
12680             1  1  
12681             0  0  
12682             1  0  
12683             0  0  

[12662 rows x 26 columns]

5. Use a bar plot to visualize the `coupon` column.

In [457]:
cleandata.groupby(['coupon','Y']).size().reset_index(name='size')

coupon  Y  size
0                    Bar  0  1190
1                    Bar  1   825
2  Carry out & Take away  0   632
3  Carry out & Take away  1  1757
4           Coffee House  0  1994
5           Coffee House  1  1995
6      Restaurant(20-50)  0   833
7      Restaurant(20-50)  1   655
8        Restaurant(<20)  0   813
9        Restaurant(<20)  1  1968

In [458]:
#bardata=cleandata.groupby('coupon').size().reset_index(name='size')
bardata=newdata.groupby(['coupon','Y']).size().reset_index(name='size')

In [459]:
fig = px.bar(bardata, x='coupon', y='size',color='Y')
fig.show()

In [460]:
newdata.groupby(['temperature','Y']).size().reset_index(name='size')

temperature  Y  size
0           30  0  1081
1           30  1  1227
2           55  0  1775
3           55  1  2058
4           80  0  2606
5           80  1  3915

6. Use a histogram to visualize the temperature column.

In [461]:
#histdata=newdata.groupby(['temperature','Y']).size().reset_index(name='size')
fig = px.histogram(newdata, x='temperature',color='Y')
fig.show()

**Investigating the Bar Coupons**

Now, we will lead you through an exploration of just the bar related coupons.  

1. Create a new `DataFrame` that contains just the bar coupons.


In [462]:
coupon = newdata.query("coupon == 'Bar'").query("Bar != 'unknown'")

In [463]:
#coupon = newdata.query("Bar != 'unknown'")

In [464]:
coupon.count()

destination             1994
passanger               1994
weather                 1994
temperature             1994
time                    1994
coupon                  1994
expiration              1994
gender                  1994
age                     1994
maritalStatus           1994
has_children            1994
education               1994
occupation              1994
income                  1994
car                     1994
Bar                     1994
CoffeeHouse             1994
CarryAway               1994
RestaurantLessThan20    1994
Restaurant20To50        1994
toCoupon_GEQ5min        1994
toCoupon_GEQ15min       1994
toCoupon_GEQ25min       1994
direction_same          1994
direction_opp           1994
Y                       1994
dtype: int64

2. What proportion of bar coupons were accepted?


In [465]:
coupon.groupby('Y').size()

Y
0    1177
1     817
dtype: int64

825 coupons are or will be accepted out of the 2015 Bar coupons (41%). 

3. Compare the acceptance rate between those who went to a bar 3 or fewer times a month to those who went more.


In [466]:
coupon.Bar.unique()

array([0, 1, 8, 4], dtype=object)

In [467]:
coupon["bar_bucket"] = pd.cut(x=coupon.Bar, bins=[-1,3,8], labels=['<3','>3'])

In [468]:
coupon.bar_bucket.unique()

['<3', '>3']
Categories (2, object): ['<3' < '>3']

In [469]:
coupon[['Bar','bar_bucket']]

Bar bar_bucket
9       0         <3
13      0         <3
17      0         <3
24      0         <3
35      0         <3
...    ..        ...
12663   0         <3
12664   0         <3
12667   0         <3
12670   0         <3
12682   0         <3

[1994 rows x 2 columns]

4. Compare the acceptance rate between drivers who go to a bar more than once a month and are over the age of 25 to the all others.  Is there a difference?


In [470]:
bardata=coupon.groupby(['bar_bucket','Y']).size().reset_index(name='size')
fig = px.bar(bardata, x='bar_bucket', y='size',color='Y')
fig.show()

C:\Users\christian.dungca\AppData\Local\Temp\ipykernel_10172\1782552677.py:1: FutureWarning:

The default of observed=False is deprecated and will be changed to True in a future version of pandas. Pass observed=False to retain current behavior or observed=True to adopt the future default and silence this warning.



In [471]:
bardata

bar_bucket  Y  size
0         <3  0  1131
1         <3  1   666
2         >3  0    46
3         >3  1   151

5. Use the same process to compare the acceptance rate between drivers who go to bars more than once a month and had passengers that were not a kid and had occupations other than farming, fishing, or forestry. 


In [472]:
# convert Bar column to int
coupon['Bar'] = coupon['Bar'].astype(int)
newcoupon = coupon.query("Bar > 1").query("passanger != 'Kid(s)'").query("occupation != 'Farming Fishing & Forestry'")

int32


In [473]:
coupon

destination  passanger weather  temperature  time coupon  \
156    No Urgent Place  Friend(s)   Sunny           80  10AM    Bar   
167               Home      Alone   Sunny           55   6PM    Bar   
171               Work      Alone   Sunny           55   7AM    Bar   
454    No Urgent Place  Friend(s)   Sunny           80  10AM    Bar   
465               Home      Alone   Sunny           55   6PM    Bar   
...                ...        ...     ...          ...   ...    ...   
12198  No Urgent Place  Friend(s)   Sunny           80  10PM    Bar   
12199  No Urgent Place  Friend(s)   Sunny           55  10PM    Bar   
12202  No Urgent Place      Alone   Rainy           55  10AM    Bar   
12207             Home    Partner   Sunny           30  10PM    Bar   
12215             Work      Alone   Snowy           30   7AM    Bar   

      expiration  gender age      maritalStatus  ...  CarryAway  \
156           1d    Male  26             Single  ...          8   
167           1d    Male  26             Single  ...          8   
171           1d    Male  26             Single  ...          8   
454           1d    Male  31             Single  ...          8   
465           1d    Male  31             Single  ...          8   
...          ...     ...  ..                ...  ...        ...   
12198         1d  Female  26  Unmarried partner  ...          1   
12199         2h  Female  26  Unmarried partner  ...          1   
12202         1d  Female  26  Unmarried partner  ...          1   
12207         2h  Female  26  Unmarried partner  ...          1   
12215         1d  Female  26  Unmarried partner  ...          1   

      RestaurantLessThan20 Restaurant20To50 toCoupon_GEQ5min  \
156                      8                8                1   
167                      8                8                1   
171                      8                8                1   
454                      8                8                1   
465                      8                8                1   
...                    ...              ...              ...   
12198                    1                1                1   
12199                    1                1                1   
12202                    1                1                1   
12207                    1                1                1   
12215                    1                1                1   

      toCoupon_GEQ15min  toCoupon_GEQ25min direction_same direction_opp  Y  \
156                   0                  0              0             1  1   
167                   0                  0              1             0  1   
171                   1                  1              0             1  1   
454                   0                  0              0             1  0   
465                   0                  0              1             0  1   
...                 ...                ...            ...           ... ..   
12198                 1                  0              0             1  1   
12199                 1                  0              0             1  1   
12202                 1                  0              0             1  1   
12207                 1                  1              0             1  1   
12215                 1                  1              0             1  1   

      bar_bucket  
156           >3  
167           >3  
171           >3  
454           >3  
465           >3  
...          ...  
12198         >3  
12199         >3  
12202         >3  
12207         >3  
12215         >3  

[189 rows x 27 columns]

In [474]:
coupon.groupby('Y').size()

Y
0     43
1    146
dtype: int64

In [476]:
coupon.maritalStatus.unique()

array(['Single', 'Unmarried partner', 'Married partner', 'Divorced'],
      dtype=object)

6. Compare the acceptance rates between those drivers who:

- go to bars more than once a month, had passengers that were not a kid, and were not widowed *OR*
- go to bars more than once a month and are under the age of 30 *OR*
- go to cheap restaurants more than 4 times a month and income is less than 50K. 



In [ ]:
# go to bars more than once a month, had passengers that were not a kid, and were not widowed
coupon['Bar'] = coupon['Bar'].astype(int)
newcoupon = coupon.query("Bar > 1").query("passanger != 'Kid(s)'").query("occupation != 'Farming Fishing & Forestry'")


In [ ]:
# go to bars more than once a month and are under the age of 30

In [ ]:
# go to cheap restaurants more than 4 times a month and income is less than 50K

7.  Based on these observations, what do you hypothesize about drivers who accepted the bar coupons?

### Independent Investigation

Using the bar coupon example as motivation, you are to explore one of the other coupon groups and try to determine the characteristics of passengers who accept the coupons.  